### PR TITLE
PathSeq documentation update

### PR DESCRIPTION
**Pull request by mwalker174**
- Moved tools to "Metagenomics" program group
- Updated tool docs
- Changed tool arguments to kebab-case
- Defined argument strings as static variables that are cross-referenced in integration tests

### CODE REVIEW

**Comment by sooheelee** [comment]
Just seeing this. I've only been looking at PRs where I am explicitly tagged within the comments (e.g. with @sooheelee). So I will take a quick look.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java [hunk 1]
```diff
@@ -28,16 +27,99 @@
...
+ *
+ * <p>Because sequence alignment is computationally burdensome, PathSeq is integrated with Apache Spark,
+ * enabling parallelization of all steps in the pipeline on multi-core workstations and cluster environments. This
+ * overcomes the high computational cost and permits rapid turnaround times (minutes to hours) in deep sequenced samples.</p>
+ *
+ * <h4>Reference files</h4>
+ *
+ * <p>Before running the PathSeq pipeline, the host and microbe references must be built. Recommended pre-built references
+ * are available as <a href="https://software.broadinstitute.org/gatk/download/bundle">resource bundles</a>.</p>
+ *
```
Comment (sooheelee) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java]: Recommended pre-built references --> Prebuilt references for a standard microbial set
are available as resource bundles --> are available in the GATK Resource Bundle.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java [hunk 2]
```diff
@@ -28,16 +27,99 @@
...
+ *   --microbe-fasta reference.fa \
+ *   --taxonomy-file taxonomy.db \
+ *   --min-clipped-read-length 60 \
+ *   --min-score-identity 0.90 \
+ *   --identity-margin 0.02 \
+ *   --scores-output scores.txt \
+ *   --output output_reads.bam \
+ *   --filter-metrics filter_metrics.txt \
+ *   --score-metrics score_metrics.txt \
+ * </pre>
```
Comment (sooheelee) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java]: See comment on example commands above

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java [hunk 3]
```diff
@@ -26,74 +25,107 @@
...
+ *
+ * <h3>Input</h3>
+ *
+ * <ul>
+ *     <li>Queryname-sorted BAM file containing only paired reads aligned to the microbe reference</li>
+ *     <li>BAM file containing only unpaired reads aligned to the microbe reference</li>
+ *     <li>*Taxonomy file generated using PathSeqBuildReferenceTaxonomy</li>
+ * </ul>
+ *
+ * <p>*Available in <a href="https://software.broadinstitute.org/gatk/download/bundle">resource bundle</a></p>
```
Comment (sooheelee) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java]: --> the GATK Resource Bundle.

**Comment by sooheelee** [comment]
I will incorporate my feedback to the branch @mwalker174.

**Comment by sooheelee** [comment]
Actually, there are many conflicts and I think I should leave rebasing to an expert. @lbergelson can you help please?

**Comment by sooheelee** [comment]
@markw will resolve conflicts that crop up after rebasing.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildKmers.java [hunk 4]
```diff
@@ -15,52 +14,130 @@
 import java.util.Collection;
 
 /**
- * SparkTool to build kmer hash set or Bloom filter from a given host reference. The output file is required by PathSeqFilterSpark.
+ * Produce a set of k-mers from the given host reference. The output file from this tool is required to run the PathSeq pipeline.
+ *
+ * <p>The tool works by scanning the reference one position at a time. It takes the k-mer (k-base subsequence) starting at each
+ * consecutive position and adds it to a set. By default, the set is stored in a hash table file.</p>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildKmers.java]: Added a link for hash table Wiki article. These aren't file formats per se, they are data structures. For example a simple list is a structure i.e.:

```
item1,item2,item3
```

The format here is comma-delimited text. I could store the same data as a list like this:

```
item1
item2
item3
```

Now it's newline-delimited text, which is a different format. The k-mer file formats are specialized to the PathSeq tools and can't be opened in any other program. I've clarified this in the doc.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildKmers.java [hunk 5]
```diff
@@ -15,52 +14,130 @@
...
+ *
+ * <p>Users also have the option to represent the k-mers set using a <a href="https://en.wikipedia.org/wiki/Bloom_filter">Bloom
+ * filter</a>. This uses less memory than the default hash set but also can produce false positives. In other words, when
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildKmers.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java [hunk 6]
```diff
@@ -21,44 +20,102 @@
...
+ *
+ * <p>The tool reads the list of sequence accessions from the given reference. For each accession, it looks up the NCBI
+ * taxonomic ID of the corresponding organism and builds a taxonomic tree containing only organisms that are
+ * represented in the reference. The reference should only contain sequences from NCBI RefSeq and/or Genbank databases.</p>
+ *
+ * <h3>Input</h3>
+ * <ul>
+ *     <li>An indexed microbe reference in FASTA format (NCBI RefSeq/Genbank sequences)</li>
+ *     <li>Downloaded NCBI RefSeq (and/or GenBank) catalog archive file(s)</li>
+ *     <li>Downloaded NCBI taxonomy dump archive file</li>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java]: I can omit it here.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java [hunk 7]
```diff
@@ -21,44 +20,102 @@
...
+ *
+ * <h3>Output</h3>
+ * <ul>
+ *     <li>A binary file containing reference taxonomy information</li>
+ * </ul>
+ *
+ * <h3>Usage examples</h3>
+ *
+ * <pre>
+ * gatk-launch PathSeqBuildKmers  \
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java [hunk 8]
```diff
@@ -21,44 +20,102 @@
...
+ *   --reference microbe_reference.fasta \
+ *   --output taxonomy.db \
+ *   --refseq-catalog RefSeq-releaseXX.catalog.gz \
+ *   --tax-dump taxdump.tar.gz \
+ *   --min-non-virus-contig-length 2000
+ * </pre>
+ *
+ * <h3>Notes</h3>
+ *
+ * <p>Often there are inconsistencies between the reference sequences, NCBI catalog, and taxonomy dump archives. To
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 9]
```diff
@@ -24,34 +22,97 @@
...
 import java.io.IOException;
 
+/**
+ * Align reads to a microbe reference using BWA-MEM and Spark. Second step in the PathSeq pipeline.
+ *
+ * <p>See PathSeqPipelineSpark for an overview of the PathSeq pipeline.</p>
+ *
+ * <p>This is a specialized version of BwaSpark designed for the PathSeq pipeline. The main difference is that only
+ * primary and alternate alignments (i.e. those appearing in the SA/XA tags) are retained, while secondary and
+ * supplementary alignments are ignored.</p>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Modified to indicate that only alignments with secondary/supplementary alignment bit flags are omitted in the output.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 10]
```diff
@@ -24,34 +22,97 @@
...
+ *
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Great will make sure I incorporate this in the tutorial.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 11]
```diff
@@ -24,34 +22,97 @@
...
+ * <h3>Input</h3>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 12]
```diff
@@ -24,34 +22,97 @@
...
+ *
+ * <ul>
+ *     <li>Unaligned queryname-sorted BAM file containing only paired reads</li>
+ *     <li>Unaligned BAM file containing only unpaired reads</li>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: At this stage in the pipeline, many paired-end reads will have their mates filtered out (e.g. due to low-quality base calls or low-complexity sequences in one but not the other). Unpaired reads can be either mate-less paired-end reads or single-end reads. I've added clarification.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 13]
```diff
@@ -24,34 +22,97 @@
...
+ *     <li>*Microbe reference BWA-MEM index image generated using BwaMemIndexImageCreator</li>
+ *     <li>*Indexed microbe reference FASTA file</li>
+ * </ul>
+ *
+ * <p>*Available in <a href="https://software.broadinstitute.org/gatk/download/bundle">resource bundle</a></p>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 14]
```diff
@@ -24,34 +22,97 @@
...
+ *
+ * <h3>Output</h3>
+ *
+ * <ul>
+ *     <li>Aligned BAM file containing the paired reads</li>
+ *     <li>Aligned BAM file containing the unpaired reads</li>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Same issue as above where it also accepts reads that were, technically, sequenced paired-end but no longer have mates in the BAM. Added clarification.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 15]
```diff
@@ -24,34 +22,97 @@
...
+ *   --paired-output output_reads_paired.bam \
+ *   --unpaired-output output_reads_unpaired.bam \
+ *   --microbe-bwa-image reference.img \
+ *   --microbe-fasta reference.fa
+ * </pre>
+ *
+ * <h3>Notes</h3>
+ *
+ * <p>For small input BAMs, it is recommended that the user reduce the BAM partition size in order to increase parallelism. Note
+ * that insert size is estimated separately in each Spark partition. Consequently partition size (and other Spark runtime
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: Yes the non-Spark example runs locally, so I added the blurb. Added Spark examples here and to the Pipeline and Filter tools (Score doesn't typically need Spark).

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java [hunk 16]
```diff
@@ -24,57 +22,97 @@
...
+ * <pre>
+ * gatk PathSeqFilterSpark  \
+ *   --input input_reads.bam \
+ *   --paired-output output_reads_paired.bam \
+ *   --unpaired-output output_reads_unpaired.bam \
+ *   --min-clipped-read-length 60 \
+ *   --kmer-file host_kmers.bfi \
+ *   --filter-bwa-image host_reference.img \
+ *   --filter-metrics metrics.txt
+ * </pre>
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java (hunk 1 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java (hunk 2 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java]: Added Spark example

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java (hunk 3 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java]: Done

**Comment by mwalker174** [comment]
@sooheelee Thank you for your review. I've addressed your suggestions. 

@droazen The branch is now also rebased on the latest master.

**Pull request merged**